lib/widgets/action_sheet.dart

```diff
@@ -162,6 +162,58 @@ class ActionSheetCancelButton extends StatelessWidget {
   }
 }
 
+/// Show a sheet of actions you can take on a channel.
+void showChannelActionSheet(BuildContext context, {
+  required int streamId,
+}) {
+  final store = PerAccountStoreWidget.of(context);
+
+  final optionButtons = <ActionSheetMenuItemButton>[];
+  final unreadCount = store.unreads.countInChannelNarrow(streamId);
+  if (unreadCount > 0) {
+    optionButtons.add(
+      MarkChannelAsReadButton(
+        streamId: streamId,
+        pageContext: context,
+      ),
+    );
+  }
+  if (optionButtons.isEmpty) {
+    // TODO(a11y): This case makes a no-op gesture handler; as a consequence,
+    //   we're presenting some UI (to people who use screen-reader software) as
+    //   though it offers a gesture interaction that it doesn't meaningfully
+    //   offer, which is confusing. The solution here is probably to remove this
+    //   is-empty case by having at least one button that's always present,
+    //   such as "copy link to channel".
+    return;
+  }
+  _showActionSheet(context, optionButtons: optionButtons);
+}
+
+class MarkChannelAsReadButton extends ActionSheetMenuItemButton {
+  const MarkChannelAsReadButton({
+    super.key,
+    required this.streamId,
+    required super.pageContext
+  });
+
+  final int streamId;
+
+  @override
+  IconData get icon => ZulipIcons.message_checked;
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    return zulipLocalizations.actionSheetOptionMarkChannelAsRead;
+  }
+
+  @override
+  void onPressed() async {
+    final narrow = ChannelNarrow(streamId);
+    await ZulipAction.markNarrowAsRead(pageContext, narrow);
+  }
```

Comment:
I see; the reasoning for `ZulipAction`, in its dartdoc, makes sense.

I see that `ZulipAction.markNarrowAsRead` takes care of giving UI feedback, so I agree a `try` / `catch` isn't needed here.

It still doesn't give one specific kind of feedback I was hoping for, though 🙂:

https://github.com/zulip/zulip-flutter/pull/1317#discussion_r1943986190
> Let's have special handling for if e is ZulipApiException, like other buttons that make API requests

Could you make that adjustment, in a separate commit, for both values of `useLegacy`? For `useLegacy` `false`, that means adjusting `updateMessageFlagsStartingFromAnchor`, which is where the UI-feedback code is in that case.